rafthttp/peer.go

```diff
@@ -15,158 +15,196 @@
+	p.reader = startStreamReader(p.dialer, &generalDecoder{}, p.recv)
+	go p.run()
+	return p
 }
 
-func (p *peer) Update(u string) {
-	p.Lock()
-	defer p.Unlock()
-	if p.stopped {
-		// TODO: not panic here?
-		panic("peer: update a stopped peer")
+func (p *peer) run() {
+	var paused bool
+	// non-blocking run loop
+	for {
+		select {
+		case m := <-p.sendc:
+			if paused {
+				continue
+			}
+			var writec chan raftpb.Message
+			var name string
+			if p.msgAppWriter.canSend(m) {
+				writec = p.msgAppWriter.msgc
+				name = "msgapp stream"
+			} else if p.writer.canSend(m) {
+				writec = p.writer.msgc
+				name = "general stream"
+			} else {
+				writec = p.pipeline.msgc
+				name = "pipeline"
+			}
+			select {
+			case writec <- m:
+			default:
+				log.Printf("peer: dropping %s because maximal number %d of %s buffer entries to %s has been reached",
+					m.Type, streamBufSize, name, p.to)
+			}
+		case mm := <-p.recvc:
+			p.msgAppDecoder.updateTerm(mm.Term)
+			if err := p.r.Process(context.TODO(), mm); err != nil {
+				log.Printf("peer: process raft message error: %v", err)
+			}
+		case u := <-p.updatec:
+			p.msgAppDialer.update(u)
+			p.dialer.update(u)
+			p.pipeline.update(u)
+		case <-p.pausec:
+			paused = true
+		case <-p.resumec:
+			paused = false
+		case <-p.stopc:
+			p.msgAppReader.stop()
+			p.msgAppWriter.stop()
+			p.reader.stop()
+			p.writer.stop()
+			p.pipeline.stop()
+			close(p.done)
+			return
+		}
 	}
-	p.u = u
-	p.pipeline.update(u)
 }
 
-// Send sends the data to the remote node. It is always non-blocking.
-// It may be fail to send data if it returns nil error.
-// TODO (xiangli): reasonable retry logic
-func (p *peer) Send(m raftpb.Message) error {
-	p.Lock()
-	defer p.Unlock()
-	if p.stopped {
-		return errors.New("peer: stopped")
-	}
-	if p.paused {
-		return nil
-	}
-
-	// move all the stream related stuff into stream
-	p.stream.invalidate(m.Term)
-	if shouldInitStream(m) && !p.stream.isOpen() {
-		u := p.u
-		// todo: steam open should not block.
-		p.stream.open(types.ID(m.From), p.id, p.cid, m.Term, p.tr, u, p.r)
-		p.batcher.Reset(time.Now())
+func (p *peer) Send(m raftpb.Message) {
+	select {
+	case p.sendc <- m:
+	case <-p.done:
+		log.Panicf("peer: unexpected stopped")
 	}
+}
 
-	var err error
-	switch {
-	case isProposal(m):
-		p.propBatcher.Batch(m)
-	case canBatch(m) && p.stream.isOpen():
-		if !p.batcher.ShouldBatch(time.Now()) {
-			err = p.pipeline.send(m)
-		}
-	case canUseStream(m):
-		if ok := p.stream.write(m); !ok {
-			err = p.pipeline.send(m)
-		}
+func (p *peer) recv(m raftpb.Message) error {
+	select {
+	case p.recvc <- m:
+		return nil
+	case <-p.done:
+		return fmt.Errorf("peer is stopped")
 	default:
-		err = p.pipeline.send(m)
+		log.Printf("peer: dropping %s because maximal number %d of receiver buffer entries from %s has been reached",
+			m.Type, recvBufSize, p.to)
+		return fmt.Errorf("recv buffer is full")
```

Comment:
@xiang90 Do you think it is reasonable to drop the received message here if the buffer is full?
The buffer size is big enough for 1 second gap.
Another solution is to let receiver wait there until the message is consumed.
When this happens, the most possible reason is that CPU is fully utilized.